File: scripts/edge_function_local_test.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
Edge Function 로컬 테스트 (Python 버전)
K-Startup API 호출 및 XML 파싱 테스트
"""

import sys
sys.stdout.reconfigure(encoding='utf-8')

import os
import requests
import xml.etree.ElementTree as ET
from datetime import datetime
import json
import re

def fetch_kstartup_api():
    """K-Startup API 호출 (Edge Function에서 사용하는 URL)"""
    print("🚀 Edge Function API 테스트 시작")
    print("="*60)
    
    # GitHub Actions에서 성공한 API URL 사용
    api_url = 'http://apis.data.go.kr/B552735/kisedKstartupService01/getAnnouncementInformation01'
    
    params = {
        'ServiceKey': 'rHwfm51FIrtIJjqRL2fJFJFvNsVEng7v7Ud0T44EKQpgKoMEJmN06LZ+KQ2wbTfW29XZSm8OzMuNCUQi+MTlsQ==',
        'pageNo': '1',
        'numOfRows': '100'
    }
    
    print(f"📡 API 호출: {api_url}")
    print(f"📋 파라미터: {params}")
    
    try:
        headers = {
            'User-Agent': 'Edge-Function-Test/1.0',
            'Accept': 'application/xml, text/xml, */*'
        }
        
        response = requests.get(api_url, params=params, headers=headers, timeout=30)
        
        print(f"📊 HTTP 상태: {response.status_code}")
        print(f"📄 응답 크기: {len(response.text)} bytes")
        
        if response.status_code == 200:
            return response.text
        else:
            print(f"❌ API 오류: {response.status_code}")
            print(f"📄 응답 내용: {response.text[:500]}...")
            return None
            
    except Exception as e:
        print(f"❌ 요청 오류: {e}")
        return None

def parse_xml_content(xml_text):
    """XML 내용 파싱 및 분석"""
    print(f"\n📄 XML 파싱 시작...")
    
    try:
        # XML 구조 분석
        print("🔍 XML 구조 분석:")
        
        # Root 요소 찾기
        if '<rss' in xml_text:
            print("  - RSS 형식 감지")
        elif '<xml' in xml_text or '<?xml' in xml_text:
            print("  - XML 형식 감지")
        else:
            print("  - HTML 또는 기타 형식")
            
        # Item 요소 찾기
        item_matches = re.findall(r'<item[^>]*>(.*?)</item>', xml_text, re.DOTALL | re.IGNORECASE)
        print(f"  - Item 요소 수: {len(item_matches)}")
        
        if len(item_matches) == 0:
            # 다른 패턴 확인
            col_matches = re.findall(r'<col[^>]*>(.*?)</col>', xml_text, re.DOTALL | re.IGNORECASE)
            print(f"  - Col 요소 수: {len(col_matches)}")
            
            data_matches = re.findall(r'<data[^>]*>(.*?)</data>', xml_text, re.DOTALL | re.IGNORECASE)
            print(f"  - Data 요소 수: {len(data_matches)}")
        
        # XML을 ElementTree로 파싱 시도
        try:
            root = ET.fromstring(xml_text)
            print(f"✅ XML 파싱 성공: Root = {root.tag}")
            
            # 모든 하위 요소 탐색
            all_elements = list(root.iter())
            print(f"📊 총 요소 수: {len(all_elements)}")
            
            # 유니크한 태그명들 출력
            unique_tags = set(elem.tag for elem in all_elements)
            print(f"🏷️ 발견된 태그들: {list(unique_tags)[:10]}")
            
            # item 또는 유사한 요소 찾기
            items = []
            
            # 다양한 패턴으로 item 찾기
            for pattern in ['item', 'data/item', './/item', 'channel/item']:
                found_items = root.findall(pattern)
                if found_items:
                    print(f"✅ '{pattern}' 패턴으로 {len(found_items)}개 아이템 발견")
                    items = found_items
                    break
            
            if not items:
                # col 패턴 확인
                cols = root.findall('.//col')
                if cols:
                    print(f"📋 Col 요소 {len(cols)}개 발견")
                    
                    # col을 item으로 그룹화
                    current_item = {}
                    items_data = []
                    
                    for col in cols:
                        name = col.get('name', '')
                        value = col.text or ''
                        
                        if name and value:
                            current_item[name] = value
                            
                            # 새로운 아이템의 시작점 감지 (예: pbanc_sn)
                            if name == 'pbanc_sn' and len(current_item) > 1:
                                items_data.append(current_item.copy())
                                current_item = {name: value}
                    
                    if current_item:
                        items_data.append(current_item)
                    
                    print(f"📦 그룹화된 아이템: {len(items_data)}개")
                    
                    # 샘플 아이템 출력
                    if items_data:
                        print("\n📋 샘플 데이터:")
                        sample = items_data[0]
                        for key, value in list(sample.items())[:5]:
                            print(f"  - {key}: {value[:50]}...")
                    
                    return items_data
            else:
                # 일반 item 요소 처리 - col 구조 확인
                items_data = []
                for item in items[:5]:  # 처음 5개만
                    item_data = {}
                    
                    # col 요소들 확인
                    cols = item.findall('col')
                    if cols:
                        print(f"🔍 Item에서 {len(cols)}개 col 발견")
                        for col in cols:
                            name = col.get('name', '')
                            value = col.text or ''
                            if name and value:
                                item_data[name] = value
                                
                        print(f"📋 Col 데이터: {list(item_data.keys())[:10]}")
                    else:
                        # 일반 자식 요소들
                        for child in item:
                            if child.text and child.text.strip():
                                item_data[child.tag] = child.text.strip()
                    
                    if item_data:
                        items_data.append(item_data)
                
                print(f"📦 처리된 아이템: {len(items_data)}개")
                
                if items_data:
                    print("\n📋 샘플 데이터:")
                    sample = items_data[0]
                    for key, value in list(sample.items())[:10]:
                        print(f"  - {key}: {str(value)[:50]}...")
                
                return items_data
                
        except ET.ParseError as e:
            print(f"❌ XML 파싱 실패: {e}")
            print("📄 XML 시작 부분:")
            print(xml_text[:500])
            return None
            
    except Exception as e:
        print(f"❌ 파싱 중 오류: {e}")
        return None

def test_data_conversion(items_data):
    """데이터 변환 테스트"""
    if not items_data:
        return
        
    print(f"\n🔧 데이터 변환 테스트...")
    
    batch_id = f"edge_test_{datetime.now().strftime('%Y%m%d_%H%M%S')}"
    
    converted_records = []
    
    for item in items_data[:3]:  # 처음 3개만 테스트
        # Edge Function 형식으로 변환 (실제 필드명 사용)
        record = {
            'collection_batch_id': batch_id,
            'source': 'k-startup-edge',
            'data_status': 'raw',
            
            # 공고 기본 정보 (실제 API 필드명)
            'pblanc_nm': item.get('biz_pbanc_nm', 'Unknown'),
            'pblanc_id': item.get('pbanc_sn', 'Unknown'),
            'pblanc_url': item.get('detl_pg_url', ''),
            
            # 기관 정보
            'organ_nm': item.get('pbanc_ntrp_nm', ''),
            'exctv_organ_nm': item.get('biz_prch_dprt_nm', ''),
            'sprv_inst': item.get('sprv_inst', ''),
            
            # 지원 대상 및 내용
            'aply_trgt_ctnt': item.get('aply_trgt_ctnt', ''),
            'bsns_sumry': item.get('pbanc_ctnt', ''),
            'supt_biz_clsfc': item.get('supt_biz_clsfc', ''),
            
            # 일정
            'pbanc_rcpt_bgng_dt': item.get('pbanc_rcpt_bgng_dt', ''),
            'pbanc_rcpt_end_dt': item.get('pbanc_rcpt_end_dt', ''),
            
            # 원본 데이터
            'raw_xml_data': item
        }
        
        converted_records.append(record)
        
        print(f"✅ 변환 완료: {record['pblanc_nm']}")
    
    print(f"\n📊 변환 결과:")
    print(f"  - 총 레코드 수: {len(converted_records)}")
    print(f"  - 배치 ID: {batch_id}")
    
    # 샘플 레코드 출력
    if converted_records:
        print(f"\n📋 샘플 레코드:")
        sample = converted_records[0]
        for key, value in sample.items():
            if key != 'raw_xml_data':
                print(f"  - {key}: {str(value)[:60]}...")

def main():
    """메인 실행 함수"""
    
    # 1. API 호출
    xml_content = fetch_kstartup_api()
    
    if not xml_content:
        print("❌ API 호출 실패")
        return
    
    # 2. XML 파싱
    items_data = parse_xml_content(xml_content)
    
    if not items_data:
        print("❌ 데이터 파싱 실패")
        return
    
    # 3. 데이터 변환 테스트
    test_data_conversion(items_data)
    
    print("\n" + "="*60)
    print("🎉 Edge Function 로컬 테스트 완료!")
    print("="*60)

if __name__ == "__main__":
    main()

File: scripts/edge_function_local_test.js
// Edge Function 로컬 테스트 버전
const https = require('https');
const http = require('http');
const fs = require('fs');

// 환경변수 직접 설정
process.env.SUPABASE_URL = 'https://csuziaogycciwgxxmahm.supabase.co';
process.env.SUPABASE_SERVICE_KEY = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImNzdXppYW9neWNjaXdneHhtYWhtIiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTc1MzYxNTc4MCwiZXhwIjoyMDY5MTkxNzgwfQ.HnhM7zSLzi7lHVPd2IVQKIACDq_YA05mBMgZbSN1c9Q';

// Supabase 클라이언트 mock (간단 버전)
class SimpleSupabaseClient {
  constructor(url, key) {
    this.url = url;
    this.key = key;
  }
  
  table(tableName) {
    return new SimpleTable(this.url, this.key, tableName);
  }
}

class SimpleTable {
  constructor(url, key, tableName) {
    this.url = url;
    this.key = key;
    this.tableName = tableName;
  }
  
  select(columns) {
    return {
      count: (type) => ({
        head: (value) => ({
          execute: async () => {
            console.log(`📊 ${this.tableName} 테이블 카운트 조회 중...`);
            return { data: 279 }; // Mock data
          }
        })
      })
    };
  }
  
  eq(column, value) {
    return {
      single: async () => {
        console.log(`🔍 중복 체크: ${column} = ${value}`);
        return { data: null }; // No duplicate found
      }
    };
  }
  
  insert(data) {
    return {
      select: (columns) => ({
        single: async () => {
          console.log(`✅ 삽입 성공: ${data.pblanc_nm || data.title || 'Unknown'}`);
          return { 
            data: { id: Math.floor(Math.random() * 10000), ...data },
            error: null
          };
        }
      })
    };
  }
}

function createClient(url, key) {
  return new SimpleSupabaseClient(url, key);
}

// HTTP 요청 함수
function fetchData(url, options = {}) {
  return new Promise((resolve, reject) => {
    const protocol = url.startsWith('https:') ? https : http;
    
    const req = protocol.request(url, {
      method: options.method || 'GET',
      headers: options.headers || {},
      timeout: options.timeout || 30000
    }, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        resolve({
          ok: res.statusCode >= 200 && res.statusCode < 300,
          status: res.statusCode,
          statusText: res.statusMessage,
          text: () => Promise.resolve(data)
        });
      });
    });
    
    req.on('error', reject);
    req.on('timeout', () => reject(new Error('Request timeout')));
    req.end();
  });
}

// XML 파싱 함수들 (Edge Function에서 복사)
function parseXmlItems(xmlText) {
  const items = [];
  
  try {
    // XML에서 item 요소 찾기
    const itemMatches = xmlText.match(/<item[^>]*>([\s\S]*?)<\/item>/gi);
    
    if (!itemMatches) {
      console.log('⚠️ XML에서 item 요소를 찾을 수 없습니다.');
      // RSS 구조 확인
      if (xmlText.includes('<rss')) {
        console.log('📄 RSS 형식 감지, 다른 구조 확인 중...');
        const channelMatch = xmlText.match(/<channel[^>]*>([\s\S]*?)<\/channel>/i);
        if (channelMatch) {
          const channelContent = channelMatch[1];
          const channelItemMatches = channelContent.match(/<item[^>]*>([\s\S]*?)<\/item>/gi);
          if (channelItemMatches) {
            return parseItemMatches(channelItemMatches);
          }
        }
      }
      return items;
    }
    
    return parseItemMatches(itemMatches);
    
  } catch (e) {
    console.error('❌ XML 파싱 오류:', e);
    return items;
  }
}

function parseItemMatches(itemMatches) {
  const items = [];
  
  itemMatches.forEach((itemXml, index) => {
    try {
      const item = {};
      
      // 각 필드 추출
      const fieldMatches = itemXml.match(/<([^>\s]+)>([\s\S]*?)<\/\1>/gi);
      
      if (fieldMatches) {
        fieldMatches.forEach(fieldXml => {
          const match = fieldXml.match(/<([^>\s]+)>([\s\S]*?)<\/\1>/i);
          if (match) {
            const [, tagName, value] = match;
            if (tagName.toLowerCase() !== 'item' && value.trim()) {
              const cleanValue = value
                .replace(/<[^>]*>/g, '')
                .replace(/&lt;/g, '<')
                .replace(/&gt;/g, '>')
                .replace(/&amp;/g, '&')
                .replace(/&quot;/g, '"')
                .replace(/&#39;/g, "'")
                .replace(/\s+/g, ' ')
                .trim();
              
              if (cleanValue) {
                item[tagName.toLowerCase()] = cleanValue;
              }
            }
          }
        });
      }
      
      if (Object.keys(item).length > 0) {
        items.push(item);
      }
    } catch (e) {
      console.log(`⚠️ 아이템 ${index + 1} 파싱 실패:`, e);
    }
  });
  
  return items;
}

function parseDate(dateStr) {
  if (!dateStr || !dateStr.trim()) return null;
  
  const cleaned = dateStr.trim();
  
  const formats = [
    /^(\d{4})-(\d{2})-(\d{2})/,
    /^(\d{4})(\d{2})(\d{2})$/,
    /^(\d{4})\.(\d{2})\.(\d{2})/,
    /^(\d{4})\/(\d{2})\/(\d{2})/,
  ];
  
  for (const regex of formats) {
    const match = cleaned.match(regex);
    if (match) {
      const [, year, month, day] = match;
      return `${year}-${month.padStart(2, '0')}-${day.padStart(2, '0')}`;
    }
  }
  
  return null;
}

// 메인 Edge Function 로직
async function runEdgeFunction() {
  const startTime = new Date();
  console.log(`🚀 Edge Function 로컬 테스트 시작 - ${startTime.toISOString()}`);

  try {
    // Supabase 클라이언트 생성
    const supabaseUrl = process.env.SUPABASE_URL;
    const supabaseKey = process.env.SUPABASE_SERVICE_KEY;
    const supabase = createClient(supabaseUrl, supabaseKey);
    
    const batchId = `local_test_${startTime.toISOString().replace(/[:.]/g, '_').slice(0, 19)}`;
    console.log(`📋 배치 ID: ${batchId}`);
    
    // 1. 기존 데이터 현황 확인
    const existingCount = await supabase
      .from('kstartup_complete')
      .select('id', { count: 'exact', head: true })
      .execute();
    
    console.log(`📊 기존 DB 데이터: ${existingCount.data || 0}개`);
    
    // 2. K-Startup API 호출 (Edge Function에서 사용하는 URL)
    const apiUrl = 'https://www.k-startup.go.kr/api/apisvc/xml/GetPblancListSvc';
    const params = new URLSearchParams({
      perPage: '100',
      page: '1',
      sortColumn: 'REG_YMD',
      sortDirection: 'DESC'
    });
    
    console.log(`📡 API 호출: ${apiUrl}`);
    
    const apiResponse = await fetchData(`${apiUrl}?${params}`, {
      headers: {
        'User-Agent': 'Local-Test/1.0'
      },
      timeout: 30000
    });
    
    if (!apiResponse.ok) {
      throw new Error(`API 호출 실패: ${apiResponse.status} ${apiResponse.statusText}`);
    }
    
    const xmlText = await apiResponse.text();
    console.log(`✅ API 응답 수신: ${xmlText.length} bytes`);
    
    // 3. XML 파싱
    const items = parseXmlItems(xmlText);
    console.log(`📄 파싱된 아이템: ${items.length}개`);
    
    if (items.length === 0) {
      console.log('⚠️ 파싱된 아이템이 없습니다.');
      console.log('📄 XML 미리보기:', xmlText.slice(0, 500));
      return;
    }
    
    // 4. 샘플 데이터 출력
    console.log('\n📋 파싱된 아이템 샘플:');
    items.slice(0, 3).forEach((item, index) => {
      console.log(`  ${index + 1}. ${item.pblancnm || item.title || 'Unknown'}`);
    });
    
    // 5. DB 저장 시뮬레이션
    let insertedCount = 0;
    let duplicateCount = 0;
    
    for (const item of items.slice(0, 5)) { // 처음 5개만 테스트
      try {
        // 중복 체크 시뮬레이션
        const existing = await supabase
          .table('kstartup_complete')
          .select('id, pblanc_nm')
          .eq('pblanc_id', item.pblancid || item.seq);
        
        if (existing.data) {
          duplicateCount++;
        } else {
          // 새 레코드 삽입 시뮬레이션
          const record = {
            collection_batch_id: batchId,
            source: 'k-startup-edge',
            pblanc_nm: item.pblancnm || item.title,
            pblanc_id: item.pblancid || item.seq,
            pblanc_url: item.pblancurl || item.link,
            organ_nm: item.organnm || item.author,
            bsns_sumry: item.bsnssumry || item.description,
            raw_xml_data: item
          };
          
          const result = await supabase
            .table('kstartup_complete')
            .insert(record)
            .select('id, pblanc_nm')
            .single();
          
          insertedCount++;
        }
      } catch (err) {
        console.error(`❌ 처리 실패: ${item.pblancnm || 'Unknown'} - ${err}`);
      }
    }
    
    const endTime = new Date();
    const duration = endTime.getTime() - startTime.getTime();
    
    console.log('\n🎉 Edge Function 테스트 완료');
    console.log(`⏱️ 소요시간: ${duration}ms`);
    console.log(`📊 결과: 삽입 ${insertedCount}개, 중복 ${duplicateCount}개`);
    
  } catch (error) {
    console.error(`❌ Edge Function 오류:`, error.message);
  }
}

// 실행
runEdgeFunction();